kava/operations_test.go

```diff
@@ -573,41 +578,339 @@ func TestFeeToOperations(t *testing.T) {
+		}
+
+		coinSums := make(map[string]*big.Int)
+
+		for _, op := range ops {
+			symbol := op.Amount.Currency.Symbol
+			value, err := types.AmountValue(op.Amount)
+			require.NoError(t, err)
+
+			sum, ok := coinSums[symbol]
+			if ok {
+				sum.Add(sum, value)
+			} else {
+				coinSums[symbol] = value
+			}
+		}
+
+		for _, sum := range coinSums {
+			assert.True(t, big.NewInt(0).Cmp(sum) == 0)
+		}
+	})
+
+	t.Run("coin operation amounts match for sender", func(t *testing.T) {
+		for _, st := range senderTracking {
+			if st.Account == nil {
+				t.Skip("no sender")
+			}
+			opCoins := sdk.NewCoins()
+			for _, op := range ops {
+				if !mustAccAddressFromBech32(op.Account.Address).Equals(st.Account) {
+					continue
+				}
+
+				symbol := op.Amount.Currency.Symbol
+				value, err := types.AmountValue(op.Amount)
+				if value.Sign() != -1 {
+					continue // exit if value is non-negative, as this is not a send
+				}
+				require.NoError(t, err)
+
+				denom, ok := Denoms[symbol]
+				require.True(t, ok)
+
+				opCoins = opCoins.Add(sdk.NewCoin(denom, sdk.NewIntFromBigInt(value.Neg(value))))
+			}
+			assert.True(t, opCoins.IsEqual(st.Balance))
+		}
+	})
+
+	t.Run("coin operation amounts match for recipient", func(t *testing.T) {
+		for _, rt := range receiverTracking {
+			if rt.Account == nil {
+				t.Skip("no recipient")
+			}
+			opCoins := sdk.NewCoins()
+			for _, op := range ops {
+				if !mustAccAddressFromBech32(op.Account.Address).Equals(rt.Account) {
+					continue
+				}
+				symbol := op.Amount.Currency.Symbol
+				value, err := types.AmountValue(op.Amount)
+				if value.Sign() != 1 {
+					continue
+				}
+				require.NoError(t, err)
+
+				denom, ok := Denoms[symbol]
+				require.True(t, ok)
+
+				opCoins = opCoins.Add(sdk.NewCoin(denom, sdk.NewIntFromBigInt(value)))
+			}
+			assert.True(t, opCoins.IsEqual(rt.Balance))
+		}
+
+	})
+
+	t.Run("each sender op has no related ops", func(t *testing.T) {
+
+		for _, op := range ops {
+			value, err := types.AmountValue(op.Amount)
+			require.NoError(t, err)
+			if value.Sign() == -1 {
+				assert.Equal(t, 0, len(op.RelatedOperations))
+			}
+		}
+	})
+
+	t.Run("each recipient op is related to a sender op", func(t *testing.T) {
+
+		for _, op := range ops {
+			value, err := types.AmountValue(op.Amount)
+			require.NoError(t, err)
+			if value.Sign() == 1 {
+				assert.Equal(t, 1, len(op.RelatedOperations))
+				relatedOpIndex := op.RelatedOperations[0].Index
+				relatedOp := ops[relatedOpIndex-ops[0].OperationIdentifier.Index]
+
+				// index matches as expected
+				assert.Equal(t, relatedOpIndex, relatedOp.OperationIdentifier.Index)
+				// values match
+				negatedRelValue, err := types.NegateValue(relatedOp.Amount.Value)
+				require.NoError(t, err)
+				assert.Equal(t, op.Amount.Value, negatedRelValue)
+
+				// currencies match
+				assert.Equal(t, op.Amount.Value, negatedRelValue)
+			}
+		}
+	})
+}
+
+func calculateCoins(ops []*types.Operation) sdk.Coins {
```

Comment:
Ditto from the other comment - good method, but best to use the log.Events to get coin amounts instead of the ops themselves.